src/transformation/visitors/switch.ts

```diff
@@ -1,61 +1,139 @@
+                            context.transformExpression(clause.expression),
+                            lua.SyntaxKind.EqualityOperator
+                        ),
+                        lua.SyntaxKind.OrOperator
+                    );
+                } else {
+                    // Next condition
+                    condition = lua.createBinaryExpression(
+                        switchVariable,
+                        context.transformExpression(clause.expression),
+                        lua.SyntaxKind.EqualityOperator
+                    );
+                }
 
-    if (concatenatedIf) {
-        statements.push(concatenatedIf);
-    }
+                // Skip empty clauses
+                if (clause.statements.length === 0) continue;
 
-    const hasDefaultCase = statement.caseBlock.clauses.some(ts.isDefaultClause);
-    statements.push(lua.createGotoStatement(`${switchName}_${hasDefaultCase ? "case_default" : "end"}`));
+                // Declare or assign condition variable
+                statements.push(
+                    isInitialCondition
+                        ? lua.createVariableDeclarationStatement(conditionVariable, condition)
+                        : lua.createAssignmentStatement(
+                              conditionVariable,
+                              lua.createBinaryExpression(conditionVariable, condition, lua.SyntaxKind.OrOperator)
+                          )
+                );
+                isInitialCondition = false;
+            } else {
+                assert(!isInitialCondition, "Default clause should never be the initial condition");
+            }
 
-    for (const [index, clause] of statement.caseBlock.clauses.entries()) {
-        const labelName = `${switchName}_case_${ts.isCaseClause(clause) ? index : "default"}`;
-        statements.push(lua.createLabelStatement(labelName));
-        statements.push(lua.createDoStatement(context.transformStatements(clause.statements)));
-    }
+            // Push if statement for case
+            statements.push(
+                lua.createIfStatement(
+                    conditionVariable,
+                    lua.createBlock(context.transformStatements(clause.statements))
+                )
+            );
+
+            // Clear condition for next clause
+            condition = undefined;
+        }
 
-    statements.push(lua.createLabelStatement(`${switchName}_end`));
+        // Amalgamate the default w/ fallthrough clauses and execute if nothing else executed above
+        const start = clauses.findIndex(c => ts.isDefaultClause(c));
+        if (start >= 0) {
+            const end = statement.caseBlock.clauses
+                .slice(start)
+                .findIndex(clause => containsBreakOrReturn([...clause.statements]));
+            const defaultStatements = statement.caseBlock.clauses
+                .slice(start, end >= 0 ? end + start + 1 : undefined)
+                .reduce<lua.Statement[]>(
```

Comment:
I see that this reduce is nice, but I would prefer if you just made defaultStatements an array and then do:
```ts
for (const { statements } of statement.caseBlock.clauses.slice(start, end >= 0 ? end + start + 1 : undefined)) {
    defaultStatements.push(...statements)
}
```
Because right now you are doing a lot of duplicate copying of arrays into each other.